art/defences/trainer/certified_trainer_pytorch.py

```diff
@@ -0,0 +1,393 @@
+
+    # if we use python 3.8 or higher use the more informative TypedDict for type hinting
+    if sys.version_info >= (3, 8):
+
+        class PGDParamDict(TypedDict):
+            """
+            A TypedDict class to define the types in the pgd_params dictionary.
+            """
+
+            eps: float
+            eps_step: float
+            max_iter: int
+            num_random_init: int
+            batch_size: int
+
+    else:
+        PGDParamDict: Dict[str, Union[int, float]]
+
+
+logger = logging.getLogger(__name__)
+
+
+class DefaultLinearScheduler:
+    """
+    Class implementing a simple linear scheduler to grow the certification radius.
+    """
+
+    def __init__(self, step_per_epoch: float, initial_bound: float = 0.0) -> None:
+        """
+        Create a .DefaultLinearScheduler instance.
+
+        :param step_per_epoch: how much to increase the certification radius every epoch
+        :param initial_bound: the initial bound to increase from
+        """
+        self.step_per_epoch = step_per_epoch
+        self.bound = initial_bound
+
+    def step(self) -> float:
+        """
+        Grow the certification radius by self.step_per_epoch
+        """
+        self.bound += self.step_per_epoch
+        return self.bound
+
+
+class AdversarialTrainerCertifiedPytorch(Trainer):
+    """
+    Class performing certified adversarial training from methods such as
+
+    | Paper link: http://proceedings.mlr.press/v80/mirman18b/mirman18b.pdf
+    | Paper link: https://arxiv.org/pdf/1810.12715.pdf
+    """
+
+    def __init__(
+        self,
+        classifier: "CERTIFIER_TYPE",
+        nb_epochs: Optional[int] = 20,
+        bound: float = 0.1,
+        loss_weighting: float = 0.1,
+        batch_size: int = 10,
+        use_certification_schedule: bool = True,
+        certification_schedule: Optional[Any] = None,
+        pgd_params: Optional["PGDParamDict"] = None,
+    ) -> None:
+        """
+        Create an :class:`.AdversarialTrainerCertified` instance.
+
+        Default values are for MNIST in pixel range 0-1.
+
+        :param classifier: Classifier to train adversarially.
+        :param pgd_params: A dictionary containing the specific parameters relating to regular PGD training.
+                           Must contain  the following keys:
+
+                               "eps": Maximum perturbation that the attacker can introduce.
+                               "eps_step": Attack step size (input variation) at each iteration.
+                               "max_iter": The maximum number of iterations.
+                               "batch_size": Size of the batch on which adversarial samples are generated.
+                               "num_random_init": Number of random initialisations within the epsilon ball.
+                                                  For num_random_init=0 starting at the original input.
+
+                           If not provided, we will default to typical MNIST values.
```

Comment:
The docs build currently fails with this param. Please change to a similar format as in https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/4ff9050441d5ff1e93aea5c5c95e8411acd4d5bf/art/attacks/inference/membership_inference/label_only_boundary_distance.py#L77